packages/lisk-p2p/src/p2p.ts

```diff
@@ -463,7 +467,10 @@ export class P2P extends EventEmitter {
 			throw new Error('Cannot start the node because it is already active');
 		}
 		await this._startPeerServer();
-		await this._startDiscovery(this._config.seedPeers);
+		// TODO: Once we will a new peer discovery then we can remove this typecasting to P2PDiscoveredPeerInfo
+		await this._startDiscovery(this._config.seedPeers as ReadonlyArray<
+			P2PDiscoveredPeerInfo
+		>);
```

Comment:
After adding the new peer discovery mechanism (e.g. in the next PR), maybe this `_startDiscovery` method could be renamed to `_startShuffling` or `_startPeerShuffling` and then we could create the new/separate discovery function/interval from scratch. Is that the intention?

In the current version of the protocol, the shuffling should only add/connect to peers (it should never intentionally remove/disconnect from peers) but once we migrate to the LIP protocol, we can adjust the shuffling function to make it remove some peers too (so each interval, it will add some peers and also remove some peers to maintain an optimum amount of peers).

Reply:
> It seems that `P2PDiscoveredPeerInfo` requires the `version` field, does this transpile (I mean the `as` typecasting)?

It does transpile with `as`, it is also missing required height but still it works but this is only temporary as I mentioned in the comments that this will be removed with the new discovery function and then we don't have to typecast. 



> After adding the new peer discovery mechanism (e.g. in the next PR), maybe this `_startDiscovery` method could be renamed to `_startShuffling` or `_startPeerShuffling` and then we could create a new/separate discovery function/interval from scratch. Is that the intention?
> 
> My thought is that the new discovery interval/algorithm should be different enough that it should be implemented from scratch as an independent interval.
> 
> In the current version of the protocol, the shuffling should only add/connect to peers (it should never intentionally remove/disconnect from peers) but once we migrate to the LIP protocol, we can adjust the shuffling function to make it remove some peers too (so each interval, it will add some peers and also remove some peers to maintain an optimum amount of peers).

Yes actually we can change it to `peerStartShuffling` but maybe when we start with phase 2 . Like you said, in current protocol it doesn;t remove any connection so it doesn;t make sense now to have it. 
What I wanted was, when we should trigger discovery again, like we do it in the start up of a node and then we should do it either at certain interval or at some trigger like if number of active connections are lower than some threshold.
I guess having both periodic and trigger is what is required, do you think we should still keep _startDiscovery periodically and using the new discovery?

Reply:
My thinking is that we shouldn't treat seed peers as special; to always combine them with the _newPeers and _triedPeers maps.

If we implement a new discovery function to discover from seed peers (who only have ip and port) then we should be able to run that exact same function on an interval against any kind of `P2PPeerInfo` including `DiscoveredPeerInfo` from the `_newPeers` and `_triedPeers`. If that discovery function can be used to cover all cases (just by invoking it on an interval), I don't see why we need a second discovery function. We do need a function to connect to discovered peers though, but it doesn't classify as discovery anymore.

Reply:
Also, from the protocol perspective, it would be ugly to have two different kinds of discovery protocols; a short one where the connection disconnects after a few milliseconds and a long one which stays open. I propose we just use the short approach; that makes the protocol more consistent, predictable and easier to debug.

Reply:
> My thinking is that we shouldn't treat seed peers as special; instead, I propose that we should combine them with the _newPeers and _triedPeers maps as input to the discovery function.
> 
> If we implement a new discovery function to discover from seed peers (who only have ip and port) then we should be able to run that exact same function on an interval against any kind of `P2PPeerInfo` including `P2PDiscoveredPeerInfo` (e.g. from the `_newPeers` and `_triedPeers` maps) and it should work exactly the same. Since the selection function is not involved in discovery, we don't need to worry about having the full `P2PDiscoveredPeerInfo`, we can just make the input type to the discovery function be `P2PPeerInfo` because all it will ever need is `wsPort` and `ipAddress`.
> 
> If that seed discovery function can be used to cover all cases and works with any kind of peer info (just by invoking it on an interval), I don't see why we need a second discovery function. We do need a function to connect to discovered peers though, but it doesn't classify as discovery anymore.

Yes I agree, having two different discovery function instead of that I will update the discovery to be more generic and accept a list of peers with Ip and Port that will be used to fetch peers and close the connection and will be independent of selection function and peer management in Peer Pool

Reply:
@jondubois I will cover it another PR (issue #1152) which is only to work on the discovery process in general. Could we merge this one if there are no more comments from your end.